include/dem/dem_action_manager.h

```diff
@@ -0,0 +1,455 @@
+ *
+ * Any event that needs some following actions are handled by the action
+ * manager. The event communicates to the action manager that it happened and
+ * the action manager will change the trigger flags to the actions to do.
+ * When this action is about to be performed, it checks with the action manager
+ * if it should be trigger.
+ * Many events can triggered the same actions, but when the action is about to
+ * be performed, it only checks if it has to be performed, not what event have
+ * happened.
+ *
+ * If you implement a new feature that needs resetting or contact search after
+ * (or other actions), you should add a new "step" function to the action
+ * manager that will set the trigger to the action to be performed.
+ * If you implement new actions, you should add a new trigger flag and a new
+ * check function to the action manager. Obviously, you should also have to
+ * initialize the trigger in the constructor, and reset it in the reset
+ * function called after a DEM iteration.
+ * If the action depends on a enabling of a feature, you should add a new
+ * function to set the enabled flag.
+ */
+class DEMActionManager
+{
+public:
+  /**
+   * @brief Copy constructor as a delete function to make sure it can not be
+   * copied. Will never be used.
+   *
+   * @param copy The object to be copied
+   */
+  DEMActionManager(const DEMActionManager &copy) = delete;
+
+  /**
+   * @brief Copy constructor as a delete function to make sure it can not be
+   * assigned. Will never be used.
+   *
+   * @param copy The object to be copied
+   */
+  DEMActionManager &
+  operator=(const DEMActionManager &copy) = delete;
+
+  /**
+   * @brief Getter of the unique instance of the DEMActionManager.
+   */
+  static DEMActionManager *
+  get_action_manager();
+
+  /**
+   * @brief Reset all triggers to false.
+   * TODO: modify the handling of the contact search at counter = 1 because of
+   * the resetting of the mobility status.
+   */
+  inline void
+  reset_triggers()
+  {
+    // First mobility status identification of the CFD time step (from the
+    // velocity computed at the first DEM time step (counter = 0) of the CFD
+    // time step) The contact search is executed to make sure the mobility
+    // status of cell match the particles that are in.
+    load_balance_trigger                 = false;
+    contact_search_trigger               = mobility_status_reset_trigger;
+    clear_tangential_overlap_trigger     = false;
+    solid_object_search_trigger          = false;
+    sparse_contacts_cells_update_trigger = false;
+    read_checkpoint_trigger              = false;
+    mobility_status_reset_trigger        = false;
+  }
+
+  /**
+   * @brief Flag that there are periodic boundary in the simulation.
+   */
+  inline void
+  set_periodic_boundaries_enabled()
+  {
+    this->periodic_boundaries_enabled = true;
+  }
+
+  /**
+   * @brief Check if the periodic boundaries are enabled to perform some actions.
```

Comment:
Periodic boundaries checker? Minor thing but "to perform some actions" is super vague :P

Reply:
It is a bit vague on purpose, because those actions are not in the class, meaning that they will probably won't be updated here even tho I write the current ones. But I did added "... that are not handled by triggers of the action manager."